File: todo-app-spring/src/main/java/com/todo/todoappspring/controller/TodoController.java


package com.todo.todoappspring.controller;

import com.todo.todoappspring.model.Todo;
import com.todo.todoappspring.service.TodoService;
import jakarta.validation.Valid; // Para usar las anotaciones de validación (NotBlank)
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/todos")
@CrossOrigin(origins = "http://localhost:4200") // Permite al frontend de Angular acceder
public class TodoController {

    @Autowired
    private TodoService todoService;

    // 1. GET - Obtener todas las tareas
    @GetMapping
    public List<Todo> getAllTodos() {
        return todoService.findAll();
    }

    // 2. GET/:id - Obtener una tarea por ID
    @GetMapping("/{id}")
    public ResponseEntity<Todo> getTodoById(@PathVariable String id) {
        return todoService.findById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    // 3. POST - Crear una nueva tarea
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Todo createTodo(@Valid @RequestBody Todo todo) {
        // @Valid activa la validación definida en la entidad (ej. @NotBlank)
        return todoService.save(todo);
    }

    // 4. PUT/:id - Actualizar una tarea existente
    @PutMapping("/{id}")
    public ResponseEntity<Todo> updateTodo(@PathVariable String id, @Valid @RequestBody Todo todoDetails) {
        try {
            Todo updatedTodo = todoService.update(id, todoDetails);
            return ResponseEntity.ok(updatedTodo);
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }

    // 5. DELETE/:id - Eliminar una tarea
    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT) // 204 No Content
    public void deleteTodo(@PathVariable String id) {
        todoService.deleteById(id);
    }
}

File: nodejs-todo-api/src/controllers/todoController.js
// src/controllers/todoController.js
const Todo = require('../models/todoModel');

// Obtener todas las tareas
exports.getAllTodos = async (req, res) => {
  try {
    const todos = await Todo.find();
    res.status(200).json(todos);
  } catch (error) {
    res.status(500).json({ message: 'Error al obtener tareas', error: error.message });
  }
};

// Crear una nueva tarea (POST)
exports.createTodo = async (req, res) => {
  try {
    const newTodo = new Todo(req.body);
    const todo = await newTodo.save();
    res.status(201).json(todo);
  } catch (error) {
    res.status(400).json({ message: 'Error al crear tarea', error: error.message });
  }
};

// Obtener una tarea por ID (GET :id)
exports.getTodoById = async (req, res) => {
  try {
    const todo = await Todo.findById(req.params.id);
    if (!todo) {
      return res.status(404).json({ message: 'Tarea no encontrada' });
    }
    res.status(200).json(todo);
  } catch (error) {
    res.status(500).json({ message: 'Error al buscar tarea', error: error.message });
  }
};

// Actualizar una tarea (PUT :id)
exports.updateTodo = async (req, res) => {
  try {
    const todo = await Todo.findByIdAndUpdate(req.params.id, req.body, { new: true, runValidators: true });
    if (!todo) {
      return res.status(404).json({ message: 'Tarea no encontrada para actualizar' });
    }
    res.status(200).json(todo);
  } catch (error) {
    res.status(400).json({ message: 'Error al actualizar tarea', error: error.message });
  }
};

// Eliminar una tarea (DELETE :id)
exports.deleteTodo = async (req, res) => {
  try {
    const todo = await Todo.findByIdAndDelete(req.params.id);
    if (!todo) {
      return res.status(404).json({ message: 'Tarea no encontrada para eliminar' });
    }
    res.status(204).send(); // 204 No Content
  } catch (error) {
    res.status(500).json({ message: 'Error al eliminar tarea', error: error.message });
  }
};